### DISCUSSION TITLE
Difference between resolvers and controllers

### DISCUSSION THREAD

**Discussion by louis1999** [graphql-java/graphql-java#2959] [Q&A] [answered]
Hello,

I'm new to spring boot and graphql so my question might be dump but I didn't see any answer on internet so i'm asking it here.
What's the difference between the graphql resolvers and the sring boot Controller ?
Most of the tutorials I see on google and youtube are using these specific ways (some are also using services but it seems less popular) of mapping a query to an execution that will retrieve data from a repository.
I can't find are the differences, do they have the same purpose or maybe I am misunderdanting something ? What is the best way to make that schema mapping between query and execution ?
I would be glad to have your thoughts about this, if my question isn't clear, tell me how i can explain it better.

Thanks for your help

**Comment by dondonz** [comment, marked as answer]
Hi @louis1999, 

In short, you're right, Spring for GraphQL's controller annotations are doing the same thing under the hood, implementing and registering a datafetcher. It's very useful and cuts out many lines of code.

By the way, the words resolver and datafetcher can be used interchangeably, in this library they happen to be called datafetchers.

In more detail: we need to add the code to fetch the data for a specific GraphQL field. Say I have a tiny schema like so (it's almost dinner time):

```graphql
type Query {
    pizzas: [String]
    iceCreams: [String]
}
```

If you're using **GraphQL Java** on its own then you need to instantiate a `RuntimeWiring` object and register every single datafetcher:

```java
RuntimeWiring.newRuntimeWiring()
	.type("Query", type -> type
		.dataFetcher("pizzas", pizzasDataFetcher())
		.dataFetcher("iceCreams", iceCreamsDataFetcher())
	)
	.build();
```

Every `DataFetcher` is a method that implements the `DataFetcher` interface. So I would have to write a separate `pizzasDataFetcher` method and `iceCreamsDataFetcher` method.

```java
private DataFetcher<List<String>> pizzasDataFetcher() {
    return dfEnv -> List.of("Supreme", "Pepperoni", "Hawaiian");
}
```

My schema is small at the moment, but if I added hundreds of extra fields, there could be hundreds more `DataFetcher`s and I'd have to register each one with `RuntimeWiring`. The field name `pizzas` is similar to the corresponding `pizzasDataFetcher` method name so there's some repetition here.

This is where **Spring for GraphQL** comes in, with the annotation based programming model, saving plenty of code.

Instead of manually instantiating `RuntimeWiring`, writing a `DataFetcher` method, and registering the `DataFetcher` method, this is how to do it in a Spring for GraphQL controller:

```java
@Controller
public class FoodController {
    @QueryMapping
    public List<String> pizzas() {
        return List.of("Supreme", "Pepperoni", "Hawaiian");
    }
}
```

Spring for GraphQL reads the `@QueryMapping` annotation, recognises there is a type called `Query` in the schema, recognises the controller method is called `pizzas` and there is a `Query` field called `pizzas`, implements a `DataFetcher` with my three pizza flavours, and registers the `DataFetcher` with the `RuntimeWiring`. All this is happening automatically. (This is only one of the many clever things Spring for GraphQL does on top of GraphQL Java.)

I have a basic list being returned in this example, but you can replace my three pizzas with a method that retrieves data from a repository, calls another service, or anything you like.

By the way, `@QueryMapping` is a shortcut annotation, meaning the field is on a `Query` type. The more general annotation is `@SchemaMapping`, which can be used for any type. More info on this in the Spring documentation https://docs.spring.io/spring-graphql/docs/current/reference/html/#controllers

And one more note, many tutorials are out of date, which is confusing. **Spring for GraphQL** went GA earlier this year, which is the official Spring integration. Many tutorials are referring to an old and unrelated project "Spring Boot GraphQL Starter" or `com.graphql-java:graphql-spring-boot-starter`, which was originally maintained by this team but is now replaced by Spring for GraphQL. To avoid confusion, look for the words "Spring for GraphQL" or `org.springframework.boot:spring-boot-starter-graphql`. In case you haven't seen it yet, we have a quick start tutorial https://www.graphql-java.com/tutorials/getting-started-with-spring-boot

I hope that helps!

P.S. It's not a dumb question :)

**Reply by louis1999** [reply]
Thanks a lot for your explanation, that clears my mind on lots of point, i will have a look on the tutorial to see if i can understand new things
